Use distance to camera in light and decal distance fade calculations

Calinou opened the pull request:
This also uses a smoother fading formula for Decal distance fade (same as OmniLight and SpotLight), superseding https://github.com/godotengine/godot/pull/58298 in the process.

I found this fix while working on a [`3.x` backport](https://github.com/Calinou/godot/tree/light-add-distance-fade-3.x) of https://github.com/godotengine/godot/pull/58512.

This closes https://github.com/godotengine/godot/issues/58513.

**Testing project:** [test_light_decal_distance_fade.zip](https://github.com/godotengine/godot/files/9403485/test_light_decal_distance_fade.zip)

Calinou committed:
Use distance to camera in light and decal distance fade calculations

This also uses a smoother fading formula for Decal distance fade
(same as OmniLight and SpotLight).

clayjohn commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -3413,45 +3413,45 @@ void RendererSceneRenderRD::_setup_lights(const PagedArray<RID> &p_lights, const
 
 				cluster.omni_light_sort[cluster.omni_light_count].instance = li;
-				cluster.omni_light_sort[cluster.omni_light_count].depth = distance;
+				cluster.omni_light_sort[cluster.omni_light_count].depth = depth;
```
Is there a compelling reason to use camera depth here instead of just using distance everywhere?

Calinou replied:
I don't know if the clustering code expects the distance to be "flat" (relative to the camera frustum). I kept it this way to avoid regressions.

clayjohn commented:
One problem we are going to face here is that we don't propogate camera information all the way through the rendering pipeline. We don't even know if the camera is perspective or orthogonal. We may have to pass more information down to the functions to make sure that we properly compute distance in both cases

Calinou commented:
Rebased and tested again, it works as expected. I added a testing project to OP as well for testing all kinds of distance fade (including GeometryInstance3D visibility range).

> We don't even know if the camera is perspective or orthogonal. 

Distance fade currently appears broken when the camera is orthogonal, but this is also an issue with GeometryInstance3D visibility range in `master`. We could try fixing all three in this PR, or do it in a separate PR.

clayjohn commented:
As discussed here and in my comment above, we need to maintain the old behaviour for orthogonal cameras. The old depth calculation is correct for orthogonal cameras. My guess is that even the clustering code should use distance to camera instead of plane, even when using a perspective camera, but that theory will need to be tested

In 3.x the code is:
```
if (p_cam_orthogonal) {
	ins->depth = near_plane.distance_to(aabb_center);
} else {
	ins->depth = p_cam_transform.origin.distance_to(aabb_center);
}
```

clayjohn commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -2957,45 +2957,45 @@ void RendererSceneRenderRD::_setup_lights(const PagedArray<RID> &p_lights, const
-							// Out of range, don't draw this light to improve performance.
-							continue;
-						}
+					if (distance > Math::pow(fade_begin + fade_length, 2)) {
```
Its probably better to do a multiply instead of a ``pow()`` call. Looking around online it sounds like GCC with -O3 optimizes pow(x, 2) to about the same as x*x but many other combinations of compiler/settings don't

clayjohn commented:
I really want to get this in before release, so are you okay if I make a PR to your branch to fix up my comments and make a few other little changes?

Calinou commented:
> I really want to get this in before release, so are you okay if I make a PR to your branch to fix up my comments and make a few other little changes?

Sure :slightly_smiling_face:

clayjohn commented:
Superseded by #71709
